File: backend/src/main/kotlin/challange/domain/Product.kt
package main.challenge.domain

import main.challenge.strategies.*

data class Product(val name: String, val type: ProductType, val price: Double)

enum class ProductType {
    PHYSICAL,
    BOOK,
    DIGITAL,
    MEMBERSHIP
}

object OrderItemProcessors {

    private val processors: HashMap<ProductType, OrderItemProcessingStrategy> = HashMap<ProductType, OrderItemProcessingStrategy>()

    fun getProcessorForProductType(productType: ProductType): OrderItemProcessingStrategy {
        if (!processors.containsKey(productType)) {
            var newProcessor: OrderItemProcessingStrategy

            when (productType) {
                ProductType.PHYSICAL -> newProcessor = PhysicalItemProcessingStrategy()
                ProductType.BOOK -> newProcessor = BookItemProcessingStrategy()
                ProductType.DIGITAL -> newProcessor = DigitalItemProcessingStrategy()
                ProductType.MEMBERSHIP -> newProcessor = MembershipItemProcessingStrategy()
            }

            processors.put(productType, newProcessor)
        }

        return processors.get(productType)!!
    }
}

File: backend/individual-assignment/typescript/src/product/product.ts
import { Order } from "../order";

export enum ProductType {
  PHYSICAL = "PHYSICAL",
  BOOK = "BOOK",
  DIGITAL = "DIGITAL",
  MEMBERSHIP = "MEMBERSHIP",
}

export interface ProductProcessorAfterPurchase {
  process(order: Order): void;
}

export class Product {
  constructor(
    public name: string,
    public type: ProductType,
    public price: number,
  ) {}
}
